src/MIDebugEngine/AD7.Impl/AD7PendingBreakpoint.cs

```diff
@@ -448,15 +460,50 @@ internal async Task DeletePendingDelete()
             {
                 bp = _bp;
                 _bp = null;
+                _pendingDelete = false;
             }
             if (bp != null)
             {
                 await bp.DeleteAsync(_engine.DebuggedProcess);
             }
         }
 
-        // Toggles the enabled state of this pending breakpoint.
-        int IDebugPendingBreakpoint2.Enable(int fEnable)
+        private int EnableUsingBindAndDelete(int fEnable)
+        {
+            bool newValue = fEnable == 0 ? false : true;
+
+            if (_enabled == newValue)
+            {
+                return Constants.S_OK;
+            }
+
+            _enabled = newValue;
+
+            // We always unbind first, even when toggling enable on, to ensure breakpoint errors show up in VS.
+            // Otherwise, a disabled breakpoint that is set before starting the debug session will not show an error when it is enabled.
+            lock (_boundBreakpoints)
+            {
+                foreach (var boundBp in _boundBreakpoints)
+                {
+                    _engine.Callback.OnBreakpointUnbound(boundBp, enum_BP_UNBOUND_REASON.BPUR_UNKNOWN);
+                }
+                (this as IDebugPendingBreakpoint2).Delete();
+                _boundBreakpoints.Clear();
+                _BPError = null;
+                 // this pending breakpoint is not actually deleted, just disabled, so override these flags
+                _deleted = false;
+                _pendingDelete = false;
```

Comment:
>_pendingDelete = false; [](start = 16, length = 23)

Does this mean that if you disable a breakpoint from run mode it will never get disabled?

Reply:
I **think** you want to leave _pendingDelete as true (if it was true in the first place), but with the following changes: 
* In [DeleteBreakpointsPendingDeletion](https://github.com/Microsoft/MIEngine/blob/d078b660535b851ab49491c05eb854e10e4b563b/src/MIDebugEngine/Engine.Impl/BreakpointManager.cs#L268-L268) change the condition to removed deleted breakpoints instead of PendingDelete breakpoints: (`b.Deleted`). It looks like this will also fix a reference leak MIEngine has since we don't otherwise remove pending breakpoints.
* Add a comment to PendingDelete to indicate that it returns true if either the breakpoint is deleted or hardware breakpoints are being used and the breakpoint is disabled.

---
In reply to: [612786532](https://github.com/microsoft/MIEngine/pull/1128#discussion_r612786532) [](ancestors = 612786532)

Reply:
Thanks for the catch. Yes, it's my mistake. Your suggestions work as intended -- I did more manual testing in run mode and issued custom `info break` commands to check that there were no GDB breakpoints being left behind anymore.